### PR TITLE
Replace the Wait_for_file_change state by a function

### PR DESCRIPTION
**Pull request by jeremiedimino**
Same deal as #5136

### CODE REVIEW

**Comment by jeremiedimino** [comment]
I also removed `Skipped_restart` as it was unused.

**Review by rgrinberg — APPROVED** [review]

**Comment by jeremiedimino** [comment]
@rgrinberg this PR includes #5132 and #5135 (that's why I left it as a draft for now). Did you review all the commits?

**Comment by jeremiedimino** [comment]
and #5136 as well.

**Comment by jeremiedimino** [comment]
@rgrinberg I reworked the code since you last reviewed as I realised there was a race with my initial version of `wait_for_chage`: if the scheduler was seeing an invalidation before `wait_for_file_change` was called, then the `wait_for_file_change` wouldn't return and might end up waiting forever. So might be worth another review.

File: src/dune_engine/scheduler.ml [hunk 1]
```diff
@@ -714,14 +714,12 @@ end = struct
...
-       filesystem events. *)
+  | (* We are not doing a build. Just accumulating invalidations until the next
+       build starts. *)
       Standing_by of
-      Memo.Invalidation.t
-  | (* Waiting for file changes to start a new a build *)
-      Waiting_for_file_changes of
-      Memo.Invalidation.t Fiber.Ivar.t
+      { invalidation : Memo.Invalidation.t
+      ; saw_insignificant_changes : bool
```
Comment (rgrinberg) [inline src/dune_engine/scheduler.ml:721]: Could you add a comment to this field?

File: src/dune_engine/scheduler.ml [hunk 2]
```diff
@@ -1134,6 +1127,23 @@ let flush_file_watcher t =
...
+    match t.status with
+    | Standing_by { invalidation; saw_insignificant_changes }
+      when (not (Memo.Invalidation.is_empty invalidation))
+           || saw_insignificant_changes ->
+      Fiber.return ()
+    | Restarting_build _ -> Fiber.return ()
+    | Standing_by _
+    | Building
+    | Shutting_down ->
+      let ivar = Fiber.Ivar.create () in
```
Comment (rgrinberg) [inline src/dune_engine/scheduler.ml:1143]: If we're shutting down, shouldn't this ivar never return?

File: src/dune_engine/scheduler.ml (hunk 1 above)
Comment (jeremiedimino) [reply src/dune_engine/scheduler.ml:721]: Sure, I added a comment

File: src/dune_engine/scheduler.ml (hunk 2 above)
Comment (jeremiedimino) [reply src/dune_engine/scheduler.ml:1148]: We could explicitly return `Fiber.never` here, but we don't do it in other functions such as `flush_file_watcher` or `wait_for_process`. It doesn't matter too much what we do here: `shutdown` sends a `Shutdown` event which will blow up the scheduler the next time we reach `Run_once.iter`.

Comment (rgrinberg) [reply src/dune_engine/scheduler.ml:1148]: I think it would be clearer if it was `never` in those places where it makes sense.

**Review by rgrinberg — APPROVED** [review]

File: src/dune_engine/scheduler.ml (hunk 2 above)
Comment (jeremiedimino) [reply src/dune_engine/scheduler.ml:1148]: That seems fine to me

**Pull request merged**